File: conversions/base64/elide-base64.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-base64

This demonstrates calling the TypeScript base64 implementation from Python
using Elide's polyglot capabilities.

Benefits:
- One base64 implementation shared across TypeScript and Python
- Consistent encoding/decoding across services
- No Python base64 library needed (or use it for validation)
- Perfect for API tokens, JWT handling, data URIs
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Python support is ready

# Assuming Elide provides something like:
# from elide import require
# base64 = require('./elide-base64.ts')

print("=== Python Calling TypeScript Base64 ===\n")

# Example 1: HTTP Basic Auth
# username = "api_user"
# password = "secret_key_123"
# credentials = f"{username}:{password}"
# encoded = base64.encode(credentials)
# print(f"Basic Auth Header: Basic {encoded}")
# print(f"Decoded: {base64.decode(encoded)}")
# print()

# Example 2: JWT Token Handling
# class JWTHandler:
#     def __init__(self):
#         self.base64 = base64
#
#     def decode_jwt_payload(self, token):
#         """Extract and decode JWT payload (middle section)"""
#         parts = token.split('.')
#         if len(parts) != 3:
#             raise ValueError("Invalid JWT format")
#
#         # Decode the payload (middle part)
#         payload = self.base64.decode(parts[1], url_safe=True)
#         return json.loads(payload)
#
# jwt_token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c"
# handler = JWTHandler()
# payload = handler.decode_jwt_payload(jwt_token)
# print(f"JWT Payload: {payload}")
# print()

# Example 3: Data URL Generation (Images)
# def create_data_url(file_path, mime_type):
#     """Convert image to base64 data URL"""
#     with open(file_path, 'rb') as f:
#         data = f.read()
#     encoded = base64.encode(data)
#     return f"data:{mime_type};base64,{encoded}"
#
# # Example usage:
# # data_url = create_data_url('logo.png', 'image/png')
# # print(f"Data URL: {data_url[:80]}...")
# # print()

# Example 4: API Token Encoding
# class APITokenService:
#     def __init__(self):
#         self.base64 = base64
#
#     def generate_token(self, user_id, secret):
#         """Generate base64-encoded API token"""
#         import time
#         timestamp = int(time.time())
#         token_data = f"{user_id}:{timestamp}:{secret}"
#         return self.base64.encode(token_data, url_safe=True)
#
#     def validate_token(self, token):
#         """Decode and validate API token"""
#         decoded = self.base64.decode(token, url_safe=True)
#         parts = decoded.split(':')
#         if len(parts) != 3:
#             return False
#         user_id, timestamp, secret = parts
#         # Validate logic here
#         return True
#
# token_service = APITokenService()
# token = token_service.generate_token("user123", "secret")
# print(f"Generated Token: {token}")
# print(f"Token Valid: {token_service.validate_token(token)}")
# print()

# Example 5: Batch Encoding for Data Pipeline
# def process_records(records):
#     """Encode sensitive fields in batch"""
#     for record in records:
#         if 'api_key' in record:
#             record['api_key_encoded'] = base64.encode(record['api_key'])
#         if 'auth_token' in record:
#             record['auth_token_encoded'] = base64.encode(record['auth_token'])
#     return records
#
# sample_records = [
#     {'user': 'alice', 'api_key': 'key_abc123'},
#     {'user': 'bob', 'api_key': 'key_xyz789'},
#     {'user': 'charlie', 'auth_token': 'token_def456'}
# ]
# processed = process_records(sample_records)
# for record in processed:
#     print(f"User: {record['user']}, Encoded: {record.get('api_key_encoded', 'N/A')}")
# print()

print("Real-world use case:")
print("- Python API encodes/decodes authentication tokens")
print("- Uses same TypeScript implementation as Node.js gateway")
print("- Guarantees consistent token format across auth services")
print("- Perfect for JWT, Basic Auth, and API tokens")
print()

print("Example: Auth Services Architecture")
print("┌─────────────────────────────────────┐")
print("│   Elide Base64 (TypeScript)        │")
print("│   conversions/base64/elide-base64  │")
print("└─────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │Gateway │          │  Auth  │")
print("    └────────┘          └────────┘")
print("         ↓                   ↓")
print("    Same token encoding everywhere!")
print()

print("Problem Solved:")
print("Before: Python base64 + JavaScript Buffer.toString('base64') = potential mismatches")
print("After: One Elide implementation = 100% consistent encoding")
print()

print("Performance Benefits:")
print("- Zero cold start overhead with Elide")
print("- Instant encoding/decoding")
print("- Shared runtime across languages")
print("- No Python library installation needed")
print()

print("When Elide Python API is ready, usage will be:")
print("  from elide import require")
print("  b64 = require('./elide-base64.ts')")
print("  encoded = b64.encode('Hello World')  # That's it!")
print("  decoded = b64.decode(encoded)")

File: conversions/base64/elide-base64.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-base64
#
# This demonstrates calling the TypeScript base64 implementation from Ruby
# using Elide's polyglot capabilities.
#
# Benefits:
# - One base64 implementation shared across TypeScript and Ruby
# - Consistent encoding/decoding across microservices
# - No Ruby Base64 module needed
# - Perfect for Rails, Sidekiq, API authentication

puts "=== Ruby Calling TypeScript Base64 ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Ruby support is ready

# Assuming Elide provides something like:
# base64 = Elide.require('./elide-base64.ts')

# Example 1: HTTP Basic Auth
# username = "api_user"
# password = "secret_key_123"
# credentials = "#{username}:#{password}"
# encoded = base64.encode(credentials)
# puts "Basic Auth Header: Basic #{encoded}"
# puts "Decoded: #{base64.decode(encoded)}"
# puts ""

# Example 2: Rails Authentication Controller
# class AuthController < ApplicationController
#   def basic_auth
#     auth_header = request.headers['Authorization']
#     if auth_header&.start_with?('Basic ')
#       encoded = auth_header.split(' ')[1]
#       credentials = base64.decode(encoded)
#       username, password = credentials.split(':')
#
#       # Authenticate user with consistent base64 decoding
#       user = User.find_by(username: username)
#       if user&.authenticate(password)
#         render json: { token: generate_jwt(user) }
#       else
#         render json: { error: 'Invalid credentials' }, status: 401
#       end
#     end
#   end
#
#   private
#   def generate_jwt(user)
#     payload = { user_id: user.id, exp: Time.now.to_i + 3600 }
#     jwt_payload = JSON.generate(payload)
#     base64.encode(jwt_payload, url_safe: true)
#   end
# end

# Example 3: Sidekiq Job with Token Encoding
# class TokenGeneratorJob
#   include Sidekiq::Worker
#
#   def perform(user_id)
#     # Generate API token using same base64 as Node.js API
#     token_data = {
#       user_id: user_id,
#       issued_at: Time.now.to_i,
#       nonce: SecureRandom.hex(16)
#     }.to_json
#
#     token = base64.encode(token_data, url_safe: true)
#
#     # Store token in Redis
#     redis.setex("api_token:#{user_id}", 3600, token)
#
#     puts "Generated token for user #{user_id}: #{token[0..20]}..."
#   end
# end

# Example 4: Data URL Generation for Images
# class ImageUploader
#   def self.to_data_url(file_path)
#     # Read file as binary
#     data = File.binread(file_path)
#
#     # Detect MIME type
#     mime_type = case File.extname(file_path).downcase
#                 when '.png' then 'image/png'
#                 when '.jpg', '.jpeg' then 'image/jpeg'
#                 when '.gif' then 'image/gif'
#                 else 'application/octet-stream'
#                 end
#
#     # Encode to base64
#     encoded = base64.encode(data)
#     "data:#{mime_type};base64,#{encoded}"
#   end
# end
#
# data_url = ImageUploader.to_data_url('avatar.png')
# puts "Data URL: #{data_url[0..80]}..."
# puts ""

# Example 5: API Token Service (Rails)
# class ApiTokenService
#   def self.generate(user_id, scope)
#     token_data = "#{user_id}:#{Time.now.to_i}:#{scope}"
#     base64.encode(token_data, url_safe: true)
#   end
#
#   def self.validate(token)
#     decoded = base64.decode(token, url_safe: true)
#     user_id, timestamp, scope = decoded.split(':')
#
#     # Check expiration (24 hours)
#     return false if Time.now.to_i - timestamp.to_i > 86400
#
#     { user_id: user_id, scope: scope }
#   rescue => e
#     false
#   end
# end
#
# token = ApiTokenService.generate(123, 'read:write')
# puts "Generated Token: #{token}"
# validation = ApiTokenService.validate(token)
# puts "Token Valid: #{validation}"
# puts ""

puts "Real-world use case:"
puts "- Ruby Rails API encodes/decodes authentication headers"
puts "- Sidekiq workers generate base64-encoded tokens"
puts "- Uses same TypeScript implementation as Node.js gateway"
puts "- Guarantees consistent encoding across all services"
puts ""

puts "Example: Microservices Architecture"
puts "┌─────────────────────────────────────┐"
puts "│   Elide Base64 (TypeScript)        │"
puts "│   conversions/base64/elide-base64  │"
puts "└─────────────────────────────────────┘"
puts "         ↓           ↓           ↓"
puts "    ┌────────┐  ┌────────┐  ┌────────┐"
puts "    │ Node.js│  │  Ruby  │  │ Rails  │"
puts "    │  API   │  │ Worker │  │  Auth  │"
puts "    └────────┘  └────────┘  └────────┘"
puts "         ↓           ↓           ↓"
puts "    Same token encoding everywhere!"
puts ""

puts "Problem Solved:"
puts "Before: Ruby Base64 + JavaScript Buffer = potential padding issues"
puts "After: One Elide implementation = 100% consistent tokens"
puts ""

puts "Rails Integration (when ready):"
puts "  # config/initializers/base64.rb"
puts "  BASE64 = Elide.require('./elide-base64.ts')"
puts "  "
puts "  # app/controllers/auth_controller.rb"
puts "  class AuthController < ApplicationController"
puts "    def basic_auth"
puts "      encoded = request.headers['Authorization'].split(' ')[1]"
puts "      credentials = BASE64.decode(encoded)"
puts "    end"
puts "  end"
puts ""

puts "Sidekiq Integration (when ready):"
puts "  class TokenJob"
puts "    include Sidekiq::Worker"
puts "    "
puts "    def perform(user_id)"
puts "      token = BASE64.encode(generate_token_data(user_id))"
puts "      # Store or send token"
puts "    end"
puts "  end"
